### PR TITLE
OpenTelemetry extension re-write using the SDK AutoConfigure

### PR DESCRIPTION
**Pull request by brunobat**
Last part of #26444 

OpenTelemetry (OTel) Extension re-write to support OTel Autoconfiguration.

**What do we get from it:**
- Support configuration according to [OTel Autoconfiguration](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md)
- People using OTel standard configs will benefit from Quarkus configuration features.
- Access to features already implemented in the autoconfig of the SDK can be easily commanded.
- Make it easier to adopt future OTel features.
- Allow using the stock OTel exporters in the future.

**What was done:**
- Documentation to comply with the changes bellow and to improve some outdated aspects like the docker collector on Jaeger.
- Cleanup and refactoring. In special, move away from the main processor instrumentation related build steps.
- Simplify sampler and support its SPI hook
- Jakarta rebase with main
- Create substitutions for the Otel Auto-configuration code using reflection but not needed for runtime
- Remove obsolete config scanning steps
- Adapt IT tests, improve their resilience and speedup their execution
- Fix additional issues from dev mode reload
- Load missing properties, fix resources and sampler
- Create Otel SDK on a bean producer
- Autoconfigure wiring
- New build and runtime configurations based on OTel guidance, ensuring backwards compatibility.

**What was left for later:**
- Optimize Configuration relocations (might not even be needed because we will drop them soon)
- TLS configs for the OTLP exporter... Was not available before anyway.
- Support additional SPI hooks for propagation, span exporter if requested by the community. 
- Fixing OTel propagation tests relying on opentelemetry-aws-xray-propagator
- Re-activate MP Telemetry Tracing using the new configs. See: https://github.com/quarkusio/quarkus/issues/31721
- **We need to consider fixing JDBC tracing later as this depends on using a later created datasource.** See: https://github.com/quarkusio/quarkus/issues/31720
- Try the `@ConfigMapping` approach

### CODE REVIEW

**Comment by brunobat** [comment]
@luneo7 Please be aware of this PR. A review from you would be much appreciated.

**Comment by radcortez** [comment]
Is this ready? It seems we have a few related test failures.

**Comment by brunobat** [comment]
At least half of the test run on a specific profile and are not executed in a standard `clean install`... Looking into it now.

**Comment by michalvavrik** [comment]
> At least half of the test run on a specific profile and are not executed in a standard `clean install`... Looking into it now.

fyi - it has to be done this way as Testcontainers for Java does not support Windows containers.

**Comment by brunobat** [comment]
Adding a wait to the test seems to have fixed the issue.

**Comment by luneo7** [comment]
Hey Bruno, I will do a review tomorrow, thanks for asking!

**Comment by brunobat** [comment]
The only remaining issue remains with the Misc4 CI tests that for some reason seem to have a broken wiring. Still evaluating them with @michalvavrik

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/DefaultConfigExtractorUtil.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        String propertyName = path + "." + member.getPropertyName();
+
+        if (member instanceof ClassDefinition.ItemMember) {
+            // leaf
+            String defaultValue = ((ClassDefinition.ItemMember) member).getDefaultValue();
+            if (defaultValue == null) {
+                return;
+            }
+            pathsAndValues.put(propertyName, defaultValue);
+            return;
```
Comment (luneo7) [inline extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/DefaultConfigExtractorUtil.java:33]: We don't actually need this return right? since we are in a if-else flow =]

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/DefaultConfigExtractorUtil.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        } else if (member instanceof ClassDefinition.GroupMember) {
+            // continue to next nodes
+            Iterable<ClassDefinition.ClassMember> members = ((ClassDefinition.GroupMember) member).getGroupDefinition()
+                    .getMembers();
+            for (ClassDefinition.ClassMember nextMember : members) {
+                extractPaths(propertyName, pathsAndValues, nextMember);
+            }
+        } else {
+            //Resource attributes
```
Comment (luneo7) [inline extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/DefaultConfigExtractorUtil.java:42]: Are we missing resource attributes implementation?

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/PropagatorType.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    PropagatorType(String value) {
+        this.value = value;
+    }
+
+    public String getValue() {
+        return value;
+    }
+
+    class Constants {
```
Comment (luneo7) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/PropagatorType.java:22]: static class?

**Comment by luneo7** [comment]
I saw some `TODO` and `FIXME` will you work on those is this PR still?

File: extensions/opentelemetry/deployment/pom.xml [hunk 4]
```diff
@@ -122,21 +122,21 @@
...
-            <groupId>io.opentelemetry.contrib</groupId>
-            <artifactId>opentelemetry-aws-resources</artifactId>
-            <scope>test</scope>
-        </dependency>
-        <dependency>
-            <groupId>io.opentelemetry.contrib</groupId>
-            <artifactId>opentelemetry-aws-xray</artifactId>
-            <scope>test</scope>
-        </dependency>
+<!--        <dependency>-->
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/deployment/pom.xml:125]: We need, however we should move these tests to a specific IT test.
The main extension should not depend on aws contrib dependencies. Current test were breaking due to changes on that lib.
Will create an issue.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/DefaultConfigExtractorUtil.java (hunk 1 above)
Comment (brunobat) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/DefaultConfigExtractorUtil.java:33]: Correct, but as I was writing it I didn't knew if I was going to write something else after the IFs.
See next comment.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/DefaultConfigExtractorUtil.java (hunk 2 above)
Comment (brunobat) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/DefaultConfigExtractorUtil.java:42]: Actually, this class is dead code. Will delete it.
In the meantime I found a better way to load all the configs.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java [hunk 5]
```diff
@@ -138,33 +147,24 @@ public void transform(TransformationContext context) {
...
-    }
+        RuntimeValue<OpenTelemetry> openTelemetry = recorder.createOpenTelemetry(shutdownContextBuildItem);
 
-    @BuildStep
-    @Record(ExecutionTime.RUNTIME_INIT)
-    void storeVertxOnContextStorage(OpenTelemetryRecorder recorder, CoreVertxBuildItem vertx) {
+        recorder.eagerlyCreateContextStorage();
         recorder.storeVertxOnContextStorage(vertx.getVertx());
+
+        instrumentationRecorder.setTracerDevMode(instrumentationRecorder.createTracers(openTelemetry));
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java:167]: good point!

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/tracing/instrumentation/InstrumentationProcessor.java [hunk 6]
```diff
@@ -102,15 +101,18 @@ VertxOptionsConsumerBuildItem vertxTracingMetricsOptions(InstrumentationRecorder
...
+            InstrumentationRecorder recorder,
             LaunchModeBuildItem launchMode) {
         Consumer<VertxOptions> vertxTracingOptions;
-        if (launchMode.getLaunchMode() == LaunchMode.DEVELOPMENT) {
-            // tracers are set in the OpenTelemetryProcessor
-            vertxTracingOptions = recorder.getVertxTracingOptionsDevMode();
-        } else {
-            vertxTracingOptions = recorder.getVertxTracingOptionsProd(recorder.createTracers());
-        }
+        //        if (launchMode.getLaunchMode() == LaunchMode.DEVELOPMENT) {
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/tracing/instrumentation/InstrumentationProcessor.java:109]: There's definitely something that got lost... We just need one of:
getVertxTracingOptionsProd
getVertxTracingOptionsDevMode
I don't think we need both any longer.

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryDisabledSdkTest.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withEmptyApplication()
+            .overrideConfigKey("quarkus.otel.sdk.disabled", "true");
+
+    @Inject
+    Tracer tracer;
+
+    @Test
+    void testNoTracer() {
+        //FIXME
```
Comment (brunobat) [inline extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/OpenTelemetryDisabledSdkTest.java:25]: This needs fixing

File: extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterConfigTest.java [hunk 8]
```diff
@@ -8,21 +8,25 @@
...
 public class OtlpExporterConfigTest {
     @RegisterExtension
     static final QuarkusUnitTest TEST = new QuarkusUnitTest()
             .withEmptyApplication()
-            .overrideConfigKey("quarkus.opentelemetry.tracer.exporter.otlp.endpoint", "http://localhost ");
+            .overrideConfigKey("otel.traces.exporter", "cdi")
+            .overrideConfigKey("otel.exporter.otlp.traces.protocol", "http/protobuf")
+            .overrideConfigKey("quarkus.opentelemetry.tracer.exporter.otlp.endpoint", "http://localhost ")
+            .overrideConfigKey("quarkus.otel.bsp.schedule.delay", "50")
+            .overrideConfigKey("quarkus.otel.bsp.export.timeout", "1s");;
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/deployment/src/test/java/io/quarkus/opentelemetry/deployment/exporter/otlp/OtlpExporterConfigTest.java:22]: will fix

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/PropagatorType.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+public enum PropagatorType {
+    TRACE_CONTEXT(Constants.TRACE_CONTEXT),
+    BAGGAGE(Constants.BAGGAGE),
+    B3(Constants.B3),
+    B3MULTI(Constants.B3MULTI),
+    JAEGER(Constants.JAEGER),
+    XRAY(Constants.XRAY),
+    OT_TRACE(Constants.OT_TRACE);
+
+    private String value;
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/PropagatorType.java:12]: makes sense

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/PropagatorType.java (hunk 3 above)
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/PropagatorType.java:22]: yes

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/SamplerType.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+public enum SamplerType {
+    ALWAYS_ON(Constants.ALWAYS_ON),
+    ALWAYS_OFF(Constants.ALWAYS_OFF),
+    TRACE_ID_RATIO(Constants.TRACE_ID_RATIO),
+    PARENT_BASED_ALWAYS_ON(Constants.PARENT_BASED_ALWAYS_ON),
+    PARENT_BASED_ALWAYS_OFF(Constants.PARENT_BASED_ALWAYS_OFF),
+    PARENT_BASED_TRACE_ID_RATIO(Constants.PARENT_BASED_TRACE_ID_RATIO);
+
+    private String value;
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/SamplerType.java:11]: yes

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/SamplerType.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+    static class Constants {
+        public static final String ALWAYS_ON = "always_on";
+        public static final String ALWAYS_OFF = "always_off";
+        public static final String TRACE_ID_RATIO = "traceidratio";
+        public static final String PARENT_BASED_ALWAYS_ON = "parentbased_always_on";
+        public static final String PARENT_BASED_ALWAYS_OFF = "parentbased_always_off";
+        public static final String PARENT_BASED_TRACE_ID_RATIO = "parentbased_traceidratio";
+    }
+}
```
Comment (brunobat) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/SamplerType.java:29]: Missing line

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtelConnectionRuntimeConfig.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+public class OtelConnectionRuntimeConfig {
+
+    // In the future this class will be reused by metrics and logs. Will hold the default properties.
+
+    public enum CompressionType {
+        GZIP("gzip"),
+        NONE("none");
+
+        private String value;
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtelConnectionRuntimeConfig.java:11]: yes

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtelConnectionRuntimeConfig.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+        CompressionType(String value) {
+            this.value = value;
+        }
+
+        public String getValue() {
+            return value;
+        }
+    }
+
+    public class Constants {
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtelConnectionRuntimeConfig.java:22]: yes

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterRuntimeConfig.java [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+    //        @ConfigItem()
+    //        public byte[] certificate;
+    //    }
+
+    /**
+     * From <a href=
+     * "https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#configuration-options">
+     * the OpenTelemetry Protocol Exporter configuration options</a>
+     */
+    public class Constants {
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterRuntimeConfig.java:92]: yes

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    public Duration timeout;
+
+    /**
+     * OTLP defines the encoding of telemetry data and the protocol used to exchange data between the client and the server.
+     * Depending on the exporter, the available protocols will be different.
+     */
+    @ConfigItem(defaultValue = Protocol.HTTP_PROTOBUF)
+    public Optional<String> protocol;
+
+    public class Protocol {
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java:84]: yes

File: integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/util/InjectionResource.java [hunk 16]
```diff
@@ -50,7 +49,7 @@ private void verifyInjections() {
         Assertions.assertNotNull(span, "Span cannot be injected");
         Assertions.assertNotNull(openTelemetry, "Baggage cannot be injected");
 
-        Assertions.assertEquals(GlobalOpenTelemetry.get(), openTelemetry);
+        //        Assertions.assertEquals(GlobalOpenTelemetry.get(), openTelemetry);
```
Comment (brunobat) [inline integration-tests/opentelemetry/src/test/java/io/quarkus/it/opentelemetry/util/InjectionResource.java:52]: duplicated

**Comment by brunobat** [comment]
@luneo7 , thanks for the review. It's appreciated.

File: extensions/opentelemetry/deployment/pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/deployment/pom.xml:125]: Was already filed.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java:167]: Was a typo. All tracers are late created now.

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/tracing/instrumentation/InstrumentationProcessor.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/tracing/instrumentation/InstrumentationProcessor.java:109]: There is no difference between dev and prod mode now as all tracers are late created.

File: extensions/opentelemetry/runtime/pom.xml [hunk 17]
```diff
@@ -44,6 +44,11 @@
             <artifactId>svm</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+        </dependency>
```
Comment (radcortez) [inline extensions/opentelemetry/runtime/pom.xml:50]: Do we really need commons-lang?

**Review by radcortez — CHANGES_REQUESTED** [review]
I would prefer that the config objects to use the @ConfigMapping approch. I know there were some initial bugs, but I think all of these are now fixed.

Let's not make a blocker for it. I can do that change later.

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/OpenTelemetryProducer.java [hunk 18]
```diff
@@ -1,18 +1,226 @@
...
+                                    if (newPropertyName.equals(QUARKUS_OTEL_TRACES_SAMPLER_PROP_NAME)) {
+                                        value = (new LegacySamplerNameConverter()).convert(value);
+                                    }
+                                    oTelConfigs.put(transformPropertyName(newPropertyName), value);
+                                }
+                            });
+                }
+            }
+        }
+        return oTelConfigs;
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/OpenTelemetryProducer.java:216]: You mean, remove the QUARKUS_OTEL_PREFIX_ENV lookups?

File: extensions/opentelemetry/runtime/pom.xml (hunk 17 above)
Comment (brunobat) [reply extensions/opentelemetry/runtime/pom.xml:50]: no, will work around it.

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/OpenTelemetryProducer.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/OpenTelemetryProducer.java:216]: Will assume yes and remove them.

File: extensions/opentelemetry/runtime/pom.xml (hunk 17 above)
Comment (brunobat) [reply extensions/opentelemetry/runtime/pom.xml]: done

File: extensions/opentelemetry/deployment/src/test/resources/resource-config/application.properties [hunk 19]
```diff
@@ -1,2 +1,5 @@
 quarkus.application.name=resource-test
-quarkus.opentelemetry.tracer.resource-attributes=service.name=authservice,service.instance.id=${quarkus.uuid}
+quarkus.otel.resource.attributes=service.name=authservice,service.instance.id=${quarkus.uuid}
+
+quarkus.otel.traces.exporter=test-span-exporter
+quarkus.otel.bsp.schedule.delay=50
```
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/deployment/src/test/resources/resource-config/application.properties:5]: done

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/OpenTelemetryProducer.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/OpenTelemetryProducer.java]: done

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java [hunk 20]
```diff
@@ -69,6 +70,14 @@ AdditionalBeanBuildItem ensureProducerIsRetained() {
                 .build();
     }
 
+    @BuildStep
+    void registerNativeImageResources(BuildProducer<ServiceProviderBuildItem> services) throws IOException {
+        services.produce(ServiceProviderBuildItem.allProvidersFromClassPath(
```
Comment (luneo7) [inline extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java:75]: Shouldn't we also add:
`io.opentelemetry.sdk.autoconfigure.spi.AutoConfigurationCustomizerProvider`
`io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider`
`io.opentelemetry.sdk.autoconfigure.spi.ConfigurablePropagatorProvider`
?

Comment (brunobat) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java:75]: I was planning to add official support for them only later. I can add the SPIs, however I'll leave testing and documentation for later because this PR is already too big.

File: docs/src/main/asciidoc/opentelemetry.adoc [hunk 21]
```diff
@@ -97,74 +108,41 @@ endpoint will be traced without any required code changes.
...
-<4> Optional gRPC headers commonly used for authentication
-<5> Add tracing information into log message.
-<6> You can also only put the trace info into the access log. In this case you must omit the info in the console log format.
+<2> gRPC endpoint to send spans
+<3> Add tracing information into log message.
+<4> You can also only put the trace info into the access log. In this case you must omit the info in the console log format.
+
+[NOTE]
+====
+All configurations have been updated from `quarkus.opentelemetry.\*` -> `quarkus.otel.*`
```
Comment (gsmet) [inline docs/src/main/asciidoc/opentelemetry.adoc:130]: Sorry to ask that but is it really an improvement? What's the rationale of this change?

File: docs/src/main/asciidoc/opentelemetry.adoc [hunk 22]
```diff
@@ -478,4 +530,14 @@ Additional exporters will be available in the Quarkiverse https://github.com/qua
...
 == OpenTelemetry Configuration Reference
 
+Quarkus supports the OpenTelemetry Autoconfiguration for Traces.
+The configurations match what you can see at
+https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md[OpenTelemetry SDK Autoconfigure]
+adding the usual `quarkus.*` prefix.
+
+Quarkus OpenTelemetry configuration properties now have the `quarkus.otel.*` prefix.
+
+*The legacy properties* with prefix `quarkus.opentelemetry.*` are currently being mapped to the new ones as a default, during a transition period. See Default column in the details bellow.
```
Comment (gsmet) [inline docs/src/main/asciidoc/opentelemetry.adoc:540]:
```suggestion
*The legacy properties* with prefix `quarkus.opentelemetry.*` are currently being mapped to the new ones as a default, during a transition period. See Default column in the details below.
```

File: extensions/opentelemetry/deployment/pom.xml (hunk 4 above)
Comment (gsmet) [inline extensions/opentelemetry/deployment/pom.xml:125]: Either you remove them entirely or if you let them commented out, please add a comment explaining why they are commented and still around. Future us will thank you :).

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/OtelBuildConfig.java [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+
+    /**
+     * Trace exporter configurations
+     */
+    public TracesBuildConfig traces;
+
+    /**
+     * No Metrics exporter for now
+     */
+    @ConfigItem(name = "metrics.exporter", defaultValue = "none")
```
Comment (gsmet) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/OtelBuildConfig.java:40]: So you will have a singleton list with `none` in it, right?

**Review by gsmet — COMMENTED** [review]
I added a few questions and very minor suggestions.

What's the status of this work exactly? Is it expected to land before our 3.0.0.CR1?

File: docs/src/main/asciidoc/opentelemetry.adoc (hunk 21 above)
Comment (brunobat) [reply docs/src/main/asciidoc/opentelemetry.adoc:130]: This is to align with the OTel side property semantics. See: https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md#disabling-opentelemetrysdk

We will transparently support them from now on.

File: docs/src/main/asciidoc/opentelemetry.adoc [hunk 24]
```diff
@@ -362,8 +334,38 @@ to the collector.
 You can set a https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#built-in-samplers[built-in sampler]
 simply by setting the desired sampler config described in the <<configuration-reference>>.
 
-If you need to use a custom sampler or to use one that is provided by one of the https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions[OpenTelemetry SDK Extensions]
-you can create a sampler producer. The OpenTelemetry extension will detect the `Sampler` CDI bean and will use it when configuring the tracer producer.
+[NOTE]
+====
+Quarkus v3.0 introduced breaking changes on the configuration.
```
[the remark this answers is not in the document]
Comment (brunobat) [reply docs/src/main/asciidoc/opentelemetry.adoc:339]: will do

File: docs/src/main/asciidoc/opentelemetry.adoc (hunk 22 above)
Comment (brunobat) [reply docs/src/main/asciidoc/opentelemetry.adoc:540]: will do

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/OtelBuildConfig.java (hunk 23 above)
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/build/OtelBuildConfig.java:40]: Yes. This will effectively disable OTel metrics.

File: docs/src/main/asciidoc/opentelemetry.adoc (hunk 21 above)
Comment (radcortez) [reply docs/src/main/asciidoc/opentelemetry.adoc:130]: When OTel introduced its configuration namespace, we discussed and decided to align with it as close as possible:
- Since the goal of OTel is to be supported across many cloud providers, tools, servers, etc, we want to stay within it. Ideally, users should use the same configuration as what they use in other external OTel aware components
- We cannot prevent users from plugin in the OTel Agent, which uses the OTel configuration as specified. This should unify that configuration, so when you plug in the agent, it will use the defined Quarkus configuration (if any)

File: extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java (hunk 20 above)
Comment (radcortez) [reply extensions/opentelemetry/deployment/src/main/java/io/quarkus/opentelemetry/deployment/OpenTelemetryProcessor.java:78]: Agree. At this point, we should move forward and add these missing pieces later. They are not a blocker at this point.

**Review by radcortez — APPROVED** [review]

**Review by luneo7 — APPROVED** [review]

File: docs/src/main/asciidoc/opentelemetry.adoc (hunk 21 above)
Comment (gsmet) [reply docs/src/main/asciidoc/opentelemetry.adoc:130]: OK. Thanks for the clarification.
I wonder if we should rewrite the properties as part of the upgrade process. Unfortunately, we cannot use regexps so we will have to map the old names to the new names manually.

Maybe it's not worth the hassle, I don't know.

**Comment by gsmet** [comment]
Let's merge. Thanks for your perseverance on this one!

**Comment by brunobat** [comment]
Thanks @gsmet. Super happy to see this merged.

**Comment by Cyberness** [comment]
How do I set my custom idGenerator now?

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/graal/Substitutions.java [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+    @TargetClass(className = "io.opentelemetry.sdk.autoconfigure.MeterProviderConfiguration")
+    static final class Target_MeterProviderConfiguration {
+
+        @Substitute
+        static List<MetricReader> configureMetricReaders(
+                ConfigProperties config,
+                ClassLoader serviceClassLoader,
+                BiFunction<? super MetricExporter, ConfigProperties, ? extends MetricExporter> metricExporterCustomizer,
+                List<Closeable> closeables) {
+            // OTel metrics not supported and there is no need to call
```
Comment (hnourhani) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/graal/Substitutions.java:30]: does it mean that we cannot use these configuration to enable Prometheus exporter:
https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/README.md#prometheus-exporter

Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/graal/Substitutions.java:30]: Correct. No OTel Metrics on Quarkus for now. It requires a lot of rework in the instrumentation.
